### PR TITLE
Integer indexed exotic object behavior

### LINKED ISSUE #374: Incorrect results for [[Get]] and [[HasProperty]] with Typed Array in prototype
There was previously an inconsistency between [[Get]]/[[Set]] and [[HasProperty]] as noted in https://github.com/tc39/ecma262/issues/347. However, we did not comply with the old semantics either (our [[HasProperty]] seemed to be guarded on the receiver as well). The new semantics remove all the receiver checks.

``` js
var o = {__proto__: new Int32Array(100)};
Object.prototype[1.3] = 100;
print(o[1.3]);
print(Reflect.has(o, '1.3'));
```

Expected: undefined/false (same as v8)
Actual: 100/true (same as sm)

### PR DESCRIPTION
**Pull request by suwc**
Implement integer indexed exotic object behaviors (9.4.5) in typed arrays:

1. Replace HasProperty(), GetProperty(), HasItem(), GetPropertyReference(), GetItem(), GetItemReference() with -Query version with tri-state return value
   TRUE  => Property_Found (1)
   FALSE => Property_NotFound (0)
         => Property_NotFound_NoProto (2) Property not found but stop going up prototype chain

2. Add new DescriptorFlags (None_NoProto) to stop going up prototype chain for
   GetSetter(), GetItemSetter()

Fixes #374

### CODE REVIEW

**Comment by suwc** [comment]
Addresses https://github.com/Microsoft/ChakraCore/issues/374

**Comment by suwc** [comment]
@dilijev @akroshg @curtisman @Microsoft/chakra-es 
Could you review please?

File: lib/Runtime/Language/JavascriptOperators.cpp [hunk 1]
```diff
@@ -1199,17 +1199,16 @@ namespace Js
...
-            result = false;
-        }
-        else
+        if (!TaggedNumber::Is(instance))
         {
             RecyclableObject* object = RecyclableObject::FromVar(instance);
-            result = object && object->GetProperty(object, propertyId, value, NULL, requestContext);
+            if (object)
+            {
+                return Property_Found == object->GetPropertyEx(object, propertyId, value, NULL, requestContext); // return false if instance is typed arrary and GetPropertyEx() returns Property_NotFound_NoProto
```
Comment (curtisman) [inline lib/Runtime/Language/JavascriptOperators.cpp]: Why does this code has to change?  Seems like the RecyclableObject::GetProperty wrapper will suffice here?

Comment (suwc) [inline lib/Runtime/Language/JavascriptOperators.cpp]: That is true.
Initially RecyclableObject::GetProperty() wrapper was checking "!= Property_NotFound", and later I determined using "== Property_Found" makes more sense. And the change here used to be necessary.
There are other occurrences of the same scenario. I will change them back. Thanks!

---
In reply to: [105235436](https://github.com/Microsoft/ChakraCore/pull/2640#discussion_r105235436) [](ancestors = 105235436)

**Comment by curtisman** [comment]
>             if (object->GetItem(instance, index, value, requestContext))

We need to change GetItem as well?

---
Refers to: lib/Runtime/Language/JavascriptOperators.cpp:3085 in 72817dd. [](commit_id = 72817dd5f462eaf0c32756c891baf06ecfafb470, deletion_comment = False)

**Comment by curtisman** [comment]
>             if (object->GetItemReference(instance, index, value, requestContext))

GetItemReference too?

---
Refers to: lib/Runtime/Language/JavascriptOperators.cpp:3104 in 72817dd. [](commit_id = 72817dd5f462eaf0c32756c891baf06ecfafb470, deletion_comment = False)

**Comment by curtisman** [comment]
>                 if (object->GetPropertyReference(instance, propertyId, value, info, requestContext))

How about GetPropertyReference?  Do we need to change this one?

---
Refers to: lib/Runtime/Language/JavascriptOperators.cpp:1992 in 72817dd. [](commit_id = 72817dd5f462eaf0c32756c891baf06ecfafb470, deletion_comment = False)

**Comment by curtisman** [comment]
>             if (object->SkipsPrototype())

I wonder if SkipsPrototype  can be used instead?  Are there cases that is not enough?

---
Refers to: lib/Runtime/Language/JavascriptOperators.cpp:1685 in 72817dd. [](commit_id = 72817dd5f462eaf0c32756c891baf06ecfafb470, deletion_comment = False)

File: lib/Runtime/Library/TypedArray.cpp [hunk 2]
```diff
@@ -606,15 +606,39 @@ namespace Js
...
 
-    BOOL TypedArrayBase::HasProperty(PropertyId propertyId)
+    BOOL TypedArrayBase::HasOwnProperty(PropertyId propertyId)
+    {
+        return HasPropertyEx(propertyId) == Property_Found;
+    }
+
+    inline BOOL TypedArrayBase::CanonicalNumericIndexString(PropertyId propertyId, ScriptContext *scriptContext)
+    {
+        PropertyString *propertyString = scriptContext->GetPropertyString(propertyId);
```
Comment (curtisman) [inline lib/Runtime/Library/TypedArray.cpp]: Seems kind of wasteful to create a property string just for this check. But I guess this is only limited to if people ask for float properties.......

Comment (suwc) [inline lib/Runtime/Library/TypedArray.cpp]: I agree. Based on perf numbers, didn't see a need for some kind of caching.

---
In reply to: [105254349](https://github.com/Microsoft/ChakraCore/pull/2640#discussion_r105254349) [](ancestors = 105254349)

**Comment by suwc** [comment]
>             if (object->SkipsPrototype())

SkipsPrototype doesn't have knowledge of propertyNameString. Only when propertyNameString is CanonicalNumericIndexString (7.1.16) would it skip prototype.

---
In reply to: [285445383](https://github.com/Microsoft/ChakraCore/pull/2640#issuecomment-285445383) [](ancestors = 285445383)

---
Refers to: lib/Runtime/Language/JavascriptOperators.cpp:1685 in 72817dd. [](commit_id = 72817dd5f462eaf0c32756c891baf06ecfafb470, deletion_comment = False)

File: lib/Runtime/Language/JavascriptOperators.cpp (hunk 1 above)
Comment (suwc) [inline lib/Runtime/Language/JavascriptOperators.cpp]: There are no other occurrences.

---
In reply to: [105237925](https://github.com/Microsoft/ChakraCore/pull/2640#discussion_r105237925) [](ancestors = 105237925,105235436)

**Comment by suwc** [comment]
>             if (object->GetItem(instance, index, value, requestContext))

No, because TypedArrayBase::GetItem() always returns true.

---
In reply to: [285441440](https://github.com/Microsoft/ChakraCore/pull/2640#issuecomment-285441440) [](ancestors = 285441440)

---
Refers to: lib/Runtime/Language/JavascriptOperators.cpp:3085 in 72817dd. [](commit_id = 72817dd5f462eaf0c32756c891baf06ecfafb470, deletion_comment = False)

**Comment by suwc** [comment]
>             if (object->GetItemReference(instance, index, value, requestContext))

There is no need for a tri-state flavor of GetItemReference, but I update TypedArrayBase::GetItemReference to catch a corner case. Also add the case to core tests.

---
In reply to: [285441519](https://github.com/Microsoft/ChakraCore/pull/2640#issuecomment-285441519) [](ancestors = 285441519)

---
Refers to: lib/Runtime/Language/JavascriptOperators.cpp:3104 in 72817dd. [](commit_id = 72817dd5f462eaf0c32756c891baf06ecfafb470, deletion_comment = False)

**Comment by suwc** [comment]
>                 if (object->GetPropertyReference(instance, propertyId, value, info, requestContext))

There is no need for a tri-state flavor of GetPropertyReference, but I update TypedArrayBase::GetItemReference to catch a corner case. Also add the case to core tests.

---
In reply to: [285441691](https://github.com/Microsoft/ChakraCore/pull/2640#issuecomment-285441691) [](ancestors = 285441691)

---
Refers to: lib/Runtime/Language/JavascriptOperators.cpp:1992 in 72817dd. [](commit_id = 72817dd5f462eaf0c32756c891baf06ecfafb470, deletion_comment = False)

**Comment by suwc** [comment]
>             if (object->GetItemReference(instance, index, value, requestContext))

No, because TypedArrayBase::GetItemReference() always returns true.

---
In reply to: [285441519](https://github.com/Microsoft/ChakraCore/pull/2640#issuecomment-285441519) [](ancestors = 285441519)

---
Refers to: lib/Runtime/Language/JavascriptOperators.cpp:3104 in 72817dd. [](commit_id = 72817dd5f462eaf0c32756c891baf06ecfafb470, deletion_comment = False)

File: lib/Runtime/Library/TypedArray.cpp (hunk 2 above)
Comment (curtisman) [inline lib/Runtime/Library/TypedArray.cpp]: Or just avoid the string creation and rewrite the CanonicalNumbericIndexString check to process a string buffer instead.
Anyways, This can be optimized if there is a need later

---
In reply to: [105298966](https://github.com/Microsoft/ChakraCore/pull/2640#discussion_r105298966) [](ancestors = 105298966,105254349)

Comment (suwc) [inline lib/Runtime/Library/TypedArray.cpp]: I took away the unnecessary type checks in JavascriptConversion::CanonicalNumericIndexString to only use the primitive operations.

---
In reply to: [105423771](https://github.com/Microsoft/ChakraCore/pull/2640#discussion_r105423771) [](ancestors = 105423771,105298966,105254349)

**Comment by suwc** [comment]
>             if (object->GetItem(instance, index, value, requestContext))

Changed for consistency among GetItem, HasItem, GetProperty, HasProperty, GetItemReference, GetPropertyReference

---
In reply to: [285521273](https://github.com/Microsoft/ChakraCore/pull/2640#issuecomment-285521273) [](ancestors = 285521273,285441440)

---
Refers to: lib/Runtime/Language/JavascriptOperators.cpp:3085 in 831dd64. [](commit_id = 831dd6419f22850bc1fc0508ae7dd7f2d58c140b, deletion_comment = False)

**Comment by suwc** [comment]
>             if (object->GetItemReference(instance, index, value, requestContext))

Changed for consistency among GetItem, HasItem, GetProperty, HasProperty, GetItemReference, GetPropertyReference

---
In reply to: [285521882](https://github.com/Microsoft/ChakraCore/pull/2640#issuecomment-285521882) [](ancestors = 285521882,285441519)

---
Refers to: lib/Runtime/Language/JavascriptOperators.cpp:3104 in 831dd64. [](commit_id = 831dd6419f22850bc1fc0508ae7dd7f2d58c140b, deletion_comment = False)

**Comment by suwc** [comment]
>             if (object->SkipsPrototype())

Changed for consistency among GetItem, HasItem, GetProperty, HasProperty, GetItemReference, GetPropertyReference

---
In reply to: [285521027](https://github.com/Microsoft/ChakraCore/pull/2640#issuecomment-285521027) [](ancestors = 285521027,285445383)

---
Refers to: lib/Runtime/Language/JavascriptOperators.cpp:1685 in 831dd64. [](commit_id = 831dd6419f22850bc1fc0508ae7dd7f2d58c140b, deletion_comment = False)

**Comment by suwc** [comment]
>                 if (object->GetPropertyReference(instance, propertyId, value, info, requestContext))

Changed for consistency among GetItem, HasItem, GetProperty, HasProperty, GetItemReference, GetPropertyReference

---
In reply to: [285521805](https://github.com/Microsoft/ChakraCore/pull/2640#issuecomment-285521805) [](ancestors = 285521805,285441691)

---
Refers to: lib/Runtime/Language/JavascriptOperators.cpp:1992 in 831dd64. [](commit_id = 831dd6419f22850bc1fc0508ae7dd7f2d58c140b, deletion_comment = False)

**Comment by suwc** [comment]
If there's no more comments I plan to merge this change by the end of tomorrow.

**Comment by curtisman** [comment]
:shipit:

**Comment by dilijev** [comment]
>     //7.1.16 CanonicalNumericIndexString(argument)

@bterlson recommends updating explicit section references to include a link to the anchor in the spec in case section numbers change in the future. In this case: [https://tc39.github.io/ecma262/#sec-canonicalnumericindexstring](https://tc39.github.io/ecma262/#sec-canonicalnumericindexstring)

---
Refers to: lib/Runtime/Language/JavascriptConversion.cpp:446 in 831dd64. [](commit_id = 831dd6419f22850bc1fc0508ae7dd7f2d58c140b, deletion_comment = False)

File: lib/Runtime/Language/JavascriptConversion.cpp [hunk 3]
```diff
@@ -450,20 +450,21 @@ namespace Js
...
+        if (JavascriptString::IsNegZero(aValue))
         {
             *indexValue = -0;
             return TRUE;
         }
-        Var indexNumberValue = JavascriptOperators::ToNumber(aValue, scriptContext);
-        if (JavascriptString::Equals(JavascriptConversion::ToString(indexNumberValue, scriptContext), aValue))
+
+        double indexDoubleValue = aValue->ToDouble();
+        if (JavascriptString::Equals(JavascriptNumber::ToStringRadix10(indexDoubleValue, scriptContext), aValue))
```
Comment (dilijev) [inline lib/Runtime/Language/JavascriptConversion.cpp]: Seems like this would be faster, but just want to check whether that's the motivation (or if there's some other motivation) behind changing from from converting String(Var)->Number(Var)->String(Var), to instead String(Var)->double(native)->String(Var). Any potential correctness concerns?

File: lib/Runtime/Language/JavascriptOperators.cpp [hunk 4]
```diff
@@ -1460,10 +1460,12 @@ namespace Js
     {
         while (JavascriptOperators::GetTypeId(instance) != TypeIds_Null)
         {
-            if (instance->HasProperty(propertyId))
+            PropertyQueryFlags result = instance->HasPropertyQuery(propertyId);
+            if (result != Property_NotFound)
             {
-                return true;
+                return JavascriptConversion::PropertyQueryFlagsToBoolean(result); // return false if instance is typed arrary and HasPropertyQuery() returns Property_Found_Undefined
```
Comment (dilijev) [inline lib/Runtime/Language/JavascriptOperators.cpp]: >arrary [](start = 119, length = 6)

nit : spelling of array

File: lib/Runtime/Types/RecyclableObject.h [hunk 5]
```diff
@@ -242,22 +249,30 @@ namespace Js {
...
         static Var DefaultEntryPoint(RecyclableObject* function, CallInfo callInfo, ...);
 
+        BOOL HasItem(uint32 index);
+        BOOL HasProperty(PropertyId propertyId);
+        BOOL GetProperty(Var originalInstance, PropertyId propertyId, Var* value, PropertyValueInfo* info, ScriptContext* requestContext);
+        BOOL GetProperty(Var originalInstance, JavascriptString* propertyNameString, Var* value, PropertyValueInfo* info, ScriptContext* requestContext);
+        BOOL GetPropertyReference(Var originalInstance, PropertyId propertyId, Var* value, PropertyValueInfo* info, ScriptContext* requestContext);
+        BOOL GetItem(Var originalInstance, uint32 index, Var* value, ScriptContext * requestContext);
+        BOOL GetItemReference(Var originalInstance, uint32 index, Var* value, ScriptContext * requestContext);
+
```
Comment (dilijev) [inline lib/Runtime/Types/RecyclableObject.h]: Are these explicitly not virtual because the intent is that they are not allowed to be overridden?

File: lib/Runtime/Language/JavascriptConversion.cpp (hunk 3 above)
Comment (suwc) [inline lib/Runtime/Language/JavascriptConversion.cpp]: Only applied following optimizations on JavascriptConversion::CanonicalNumericIndexString():
- Type specialization (remove branches that are not applicable)
- Cancellation of boxing and unboxing (Number(Var) vs. double)
- Inlining
Equivalency is guaranteed.

---
In reply to: [106289593](https://github.com/Microsoft/ChakraCore/pull/2640#discussion_r106289593) [](ancestors = 106289593)

File: lib/Runtime/Types/RecyclableObject.h (hunk 5 above)
Comment (suwc) [inline lib/Runtime/Types/RecyclableObject.h]: Yes, that is the intent. Instead override -Query version.

---
In reply to: [106294184](https://github.com/Microsoft/ChakraCore/pull/2640#discussion_r106294184) [](ancestors = 106294184)

File: lib/Runtime/Language/JavascriptOperators.cpp (hunk 4 above)
Comment (suwc) [inline lib/Runtime/Language/JavascriptOperators.cpp]: Good catch.

---
In reply to: [106293996](https://github.com/Microsoft/ChakraCore/pull/2640#discussion_r106293996) [](ancestors = 106293996)

**Comment by suwc** [comment]
>     //7.1.16 CanonicalNumericIndexString(argument)

Will change.

---
In reply to: [286886221](https://github.com/Microsoft/ChakraCore/pull/2640#issuecomment-286886221) [](ancestors = 286886221)

---
Refers to: lib/Runtime/Language/JavascriptConversion.cpp:446 in 831dd64. [](commit_id = 831dd6419f22850bc1fc0508ae7dd7f2d58c140b, deletion_comment = False)

File: lib/Runtime/Language/JavascriptConversion.h [hunk 6]
```diff
@@ -12,10 +12,13 @@ namespace Js {
         static Var OrdinaryToPrimitive(Var aValue, JavascriptHint hint, ScriptContext * scriptContext);
         static Var MethodCallToPrimitive(Var aValue, JavascriptHint hint, ScriptContext * scriptContext);
         static Var ToPrimitive(Var aValue, JavascriptHint hint, ScriptContext * scriptContext);
-        static BOOL CanonicalNumericIndexString(Var aValue, double *indexValue, ScriptContext * scriptContext);
+        static BOOL CanonicalNumericIndexString(JavascriptString *aValue, double *indexValue, ScriptContext * scriptContext);
 
         static void ToPropertyKey(Var argument, ScriptContext* scriptContext, const PropertyRecord** propertyRecord);
 
+        static PropertyQueryFlags BooleanToPropertyQueryFlags(BOOL val) { return val ? Property_Found : Property_NotFound; }
+        static BOOL PropertyQueryFlagsToBoolean(PropertyQueryFlags val) { return val == Property_Found; }
```
Comment (dilijev) [inline lib/Runtime/Language/JavascriptConversion.h]: >BOOL PropertyQueryFlagsToBoolean [](start = 15, length = 32)

minor naming nit: (semantically this is fine) -- but instead of ToBoolean perhaps this should be ToBOOL for the same reasons that we make a distinction between BOOL and bool for various API return types. It seems that the types involved here are all BOOL and it might be good to capture that.

Although, IMO this name looks and reads better.

File: lib/Runtime/Types/ActivationObject.cpp [hunk 7]
```diff
@@ -257,24 +257,24 @@ namespace Js
...
     {
-        if (__super::GetProperty(originalInstance, propertyId, value, info, requestContext))
+        if (JavascriptConversion::PropertyQueryFlagsToBoolean(__super::GetPropertyQuery(originalInstance, propertyId, value, info, requestContext)))
         {
             GetPropertyCore(info, requestContext);
-            return TRUE;
+            return Property_Found;
         }
-        return FALSE;
+        return Property_NotFound;
```
Comment (dilijev) [inline lib/Runtime/Types/ActivationObject.cpp]: >Property_NotFound [](start = 15, length = 17)

Do we not care about the potential Property_NotFound_NoProto in these cases?

File: lib/Runtime/Language/JavascriptConversion.cpp (hunk 3 above)
Comment (dilijev) [inline lib/Runtime/Language/JavascriptConversion.cpp]: Thanks

---
In reply to: [106295570](https://github.com/Microsoft/ChakraCore/pull/2640#discussion_r106295570) [](ancestors = 106295570,106289593)

File: lib/Runtime/Language/JavascriptConversion.h (hunk 6 above)
Comment (suwc) [inline lib/Runtime/Language/JavascriptConversion.h]: This is following line 27-30 below. Seems like "ToBoolean" => BOOL, and "ToBool" => bool.

---
In reply to: [106296732](https://github.com/Microsoft/ChakraCore/pull/2640#discussion_r106296732) [](ancestors = 106296732)

File: lib/Runtime/Types/ActivationObject.cpp (hunk 7 above)
Comment (suwc) [inline lib/Runtime/Types/ActivationObject.cpp]: No, because TypedArrayBase is not in ActivationObjectEx's inheritance chain.

---
In reply to: [106297091](https://github.com/Microsoft/ChakraCore/pull/2640#discussion_r106297091) [](ancestors = 106297091)

Comment (dilijev) [inline lib/Runtime/Types/ActivationObject.cpp]: Got it, thanks.

---
In reply to: [106298387](https://github.com/Microsoft/ChakraCore/pull/2640#discussion_r106298387) [](ancestors = 106298387,106297091)

File: lib/Runtime/Library/TypedArray.cpp [hunk 8]
```diff
@@ -606,15 +606,39 @@ namespace Js
...
+        ScriptContext *scriptContext = GetScriptContext();
+        if (scriptContext->IsNumericPropertyId(propertyId, &index))
         {
             // All the slots within the length of the array are valid.
-            return true;
+            return index < this->GetLength() ? Property_Found : Property_NotFound_NoProto;
         }
-        return DynamicObject::HasProperty(propertyId);
+
+        if (!scriptContext->GetPropertyName(propertyId)->IsSymbol() && CanonicalNumericIndexString(propertyId, scriptContext))
```
Comment (dilijev) [inline lib/Runtime/Library/TypedArray.cpp]: >!scriptContext->GetPropertyName(propertyId)->IsSymbol() [](start = 12, length = 55)

I'm a little bit confused here. Is it possible for a symbol to be numeric?

File: lib/Runtime/Library/TypedArray.cpp [hunk 9]
```diff
@@ -606,15 +606,39 @@ namespace Js
...
+        double result;
+        return JavascriptConversion::CanonicalNumericIndexString(propertyString, &result, scriptContext);
+    }
+
+    PropertyQueryFlags TypedArrayBase::HasPropertyQuery(PropertyId propertyId)
     {
         uint32 index = 0;
-        if (GetScriptContext()->IsNumericPropertyId(propertyId, &index) && (index < this->GetLength()))
+        ScriptContext *scriptContext = GetScriptContext();
+        if (scriptContext->IsNumericPropertyId(propertyId, &index))
```
Comment (dilijev) [inline lib/Runtime/Library/TypedArray.cpp]: >IsNumericPropertyId [](start = 27, length = 19)

Does IsNumericPropertyId also set the index if it is an integer? This seems like a poorly-named method.

File: lib/Runtime/Library/TypedArray.cpp [hunk 10]
```diff
@@ -606,15 +606,39 @@ namespace Js
...
         {
             // All the slots within the length of the array are valid.
-            return true;
+            return index < this->GetLength() ? Property_Found : Property_NotFound_NoProto;
```
Comment (dilijev) [inline lib/Runtime/Library/TypedArray.cpp]: >index < this->GetLength() ? Property_Found [](start = 19, length = 42)

I believe it is only Property_Found if it is an *integer* numeric property. Anything which is not an integer is also Property_NotFound_NoProto

File: lib/Runtime/Library/TypedArray.cpp (hunk 8 above)
Comment (suwc) [inline lib/Runtime/Library/TypedArray.cpp]: Yes. For example: 
var p = new Symbol('12');
p will have propertyString '12' and be considered numeric by CanonicalNumericIndexString().

Comment (dilijev) [inline lib/Runtime/Library/TypedArray.cpp]: Okay, makes sense. I was thinking about the built-in symbols.

---
In reply to: [106300532](https://github.com/Microsoft/ChakraCore/pull/2640#discussion_r106300532) [](ancestors = 106300532)

File: lib/Runtime/Library/TypedArray.cpp [hunk 11]
```diff
@@ -678,32 +714,39 @@ namespace Js
...
-    BOOL TypedArrayBase::GetItem(Var originalInstance, uint32 index, Var* value, ScriptContext* requestContext)
+    PropertyQueryFlags TypedArrayBase::GetItemQuery(Var originalInstance, uint32 index, Var* value, ScriptContext* requestContext)
     {
         *value = DirectGetItem(index);
-        return true;
+        return index < GetLength() ? Property_Found : Property_NotFound_NoProto;
     }
 
-    BOOL TypedArrayBase::GetItemReference(Var originalInstance, uint32 index, Var* value, ScriptContext* requestContext)
+    PropertyQueryFlags TypedArrayBase::GetItemReferenceQuery(Var originalInstance, uint32 index, Var* value, ScriptContext* requestContext)
```
Comment (dilijev) [inline lib/Runtime/Library/TypedArray.cpp]: >GetItemReferenceQuery [](start = 39, length = 21)

What's the difference between GetItemQuery and GetItemReferenceQuery?

File: lib/Runtime/Library/TypedArray.cpp [hunk 12]
```diff
@@ -715,10 +758,49 @@ namespace Js
         AssertMsg(!PropertyRecord::IsPropertyNameNumeric(propertyNameString->GetString(), propertyNameString->GetLength()),
             "Numeric property names should have been converted to uint or PropertyRecord*");
 
+        ScriptContext *requestContext = GetScriptContext();
+        if (CanonicalNumericIndexString(propertyNameString, requestContext))
+        {
+            return FALSE;
```
Comment (dilijev) [inline lib/Runtime/Library/TypedArray.cpp]: >return FALSE; [](start = 12, length = 13)

This is to cover the assert case in release builds?

File: lib/Runtime/Library/TypedArray.cpp [hunk 13]
```diff
@@ -715,10 +758,49 @@ namespace Js
...
+    {
+        if (CanonicalNumericIndexString(propertyNameString, requestContext))
+        {
+            return None_NoProto;
+        }
+
+        return __super::GetSetter(propertyNameString, setterValue, info, requestContext);
+    }
+
+
```
Comment (dilijev) [inline lib/Runtime/Library/TypedArray.cpp]: nit : extra blank line

File: lib/Runtime/Library/TypedArray.cpp (hunk 9 above)
Comment (suwc) [inline lib/Runtime/Library/TypedArray.cpp]: I can understand this naming can be confusing.

File: lib/Runtime/Library/TypedArray.cpp (hunk 10 above)
Comment (suwc) [inline lib/Runtime/Library/TypedArray.cpp]: The integer checking is done by scriptContext->IsNumericPropertyId()

File: lib/Runtime/Library/ModuleRoot.cpp [hunk 14]
```diff
@@ -406,29 +406,29 @@ namespace Js
         return this->GetLibrary()->GetGlobalObject()->GlobalObject::DeleteRootProperty(propertyId, flags);
     }
 
-    BOOL ModuleRoot::HasItem(uint32 index)
+    PropertyQueryFlags ModuleRoot::HasItemQuery(uint32 index)
     {
-        return DynamicObject::HasItem(index)
-            || (this->hostObject && JavascriptOperators::HasItem(this->hostObject, index));
+        return JavascriptConversion::BooleanToPropertyQueryFlags(JavascriptConversion::PropertyQueryFlagsToBoolean((DynamicObject::HasItemQuery(index)))
```
Comment (dilijev) [inline lib/Runtime/Library/ModuleRoot.cpp]: >( [](start = 115, length = 1)

nit : extra parens around this expression

Comment (dilijev) [inline lib/Runtime/Library/ModuleRoot.cpp]: >JavascriptConversion::PropertyQueryFlagsToBoolean [](start = 65, length = 49)

Seems like an unnecessary conversion here, maybe BooleanToPropertyQueryFlags can just handle this natively since Property_NotFound == FALSE == 0 and everything else is TRUE

File: lib/Runtime/Library/GlobalObject.cpp [hunk 15]
```diff
@@ -1748,11 +1748,11 @@ namespace Js
         return FALSE;
     }
 
-    BOOL GlobalObject::HasProperty(PropertyId propertyId)
+    PropertyQueryFlags GlobalObject::HasPropertyQuery(PropertyId propertyId)
     {
-        return DynamicObject::HasProperty(propertyId) ||
+        return JavascriptConversion::BooleanToPropertyQueryFlags(JavascriptConversion::PropertyQueryFlagsToBoolean(DynamicObject::HasPropertyQuery(propertyId)) ||
```
Comment (dilijev) [inline lib/Runtime/Library/GlobalObject.cpp]: >JavascriptConversion::PropertyQueryFlagsToBoolean [](start = 65, length = 49)

Similar to other comment -- is this conversion unnecessary?

File: lib/Runtime/Library/GlobalObject.cpp [hunk 16]
```diff
@@ -2175,27 +2175,28 @@ namespace Js
         return TRUE;
     }
 
-    BOOL GlobalObject::HasItem(uint32 index)
+    PropertyQueryFlags GlobalObject::HasItemQuery(uint32 index)
     {
-        return DynamicObject::HasItem(index)
+        return JavascriptConversion::BooleanToPropertyQueryFlags(JavascriptConversion::PropertyQueryFlagsToBoolean((DynamicObject::HasItemQuery(index)))
```
Comment (dilijev) [inline lib/Runtime/Library/GlobalObject.cpp]: >JavascriptConversion::PropertyQueryFlagsToBoolean((DynamicObject::HasItemQuery(index))) [](start = 65, length = 87)

similar to other comment : is this conversion unnecessary?

File: lib/Runtime/Library/GlobalObject.cpp [hunk 17]
```diff
@@ -2175,27 +2175,28 @@ namespace Js
...
-        if (DynamicObject::GetItemReference(originalInstance, index, value, requestContext))
+        if (JavascriptConversion::PropertyQueryFlagsToBoolean(DynamicObject::GetItemReferenceQuery(originalInstance, index, value, requestContext)))
         {
-            return TRUE;
+            return Property_Found;
         }
-        return (this->directHostObject && this->directHostObject->GetItemReference(originalInstance, index, value, requestContext)) ||
-            (this->hostObject && this->hostObject->GetItemReference(originalInstance, index, value, requestContext));
+        return JavascriptConversion::BooleanToPropertyQueryFlags(
+            (this->directHostObject && JavascriptConversion::PropertyQueryFlagsToBoolean(this->directHostObject->GetItemReferenceQuery(originalInstance, index, value, requestContext))) ||
```
Comment (dilijev) [inline lib/Runtime/Library/GlobalObject.cpp]: >JavascriptConversion::PropertyQueryFlagsToBoolean [](start = 39, length = 49)

same, is this conversion necessary?

**Comment by dilijev** [comment]
:shipit:

File: lib/Runtime/Language/JavascriptConversion.h (hunk 6 above)
Comment (dilijev) [inline lib/Runtime/Language/JavascriptConversion.h]: Sounds good

File: lib/Runtime/Library/TypedArray.cpp (hunk 10 above)
Comment (dilijev) [inline lib/Runtime/Library/TypedArray.cpp]: I see... Even more reason that method it poorly-named :P Maybe it should be called TryGetIntegralPropertyId

Comment (suwc) [inline lib/Runtime/Library/TypedArray.cpp]: Good idea, although I found 131 occurrences in 24 files. Since this is not a new method name introduced in this changeset, I would leave it for future cosmetic PRs.

---
In reply to: [106311219](https://github.com/Microsoft/ChakraCore/pull/2640#discussion_r106311219) [](ancestors = 106311219)

File: lib/Runtime/Library/GlobalObject.cpp (hunk 17 above)
Comment (suwc) [inline lib/Runtime/Library/GlobalObject.cpp]: Yes, because the definition of JavascriptConversion::PropertyQueryFlagsToBoolean() can be easily changed in the future without touching all these sites.


---
In reply to: [106305716](https://github.com/Microsoft/ChakraCore/pull/2640#discussion_r106305716) [](ancestors = 106305716)

File: lib/Runtime/Library/GlobalObject.cpp (hunk 16 above)
Comment (suwc) [inline lib/Runtime/Library/GlobalObject.cpp]: Yes - convert to BOOL for logic operations and back to PropertyQueryFlags to satisfy compiler type-checking.
These long-name converters encapsulates logic that can be updated easily in the future.

---
In reply to: [106305680](https://github.com/Microsoft/ChakraCore/pull/2640#discussion_r106305680) [](ancestors = 106305680)

File: lib/Runtime/Library/GlobalObject.cpp (hunk 15 above)
Comment (suwc) [inline lib/Runtime/Library/GlobalObject.cpp]: Yes - see other comment.

---
In reply to: [106305579](https://github.com/Microsoft/ChakraCore/pull/2640#discussion_r106305579) [](ancestors = 106305579)

File: lib/Runtime/Library/ModuleRoot.cpp (hunk 14 above)
Comment (suwc) [inline lib/Runtime/Library/ModuleRoot.cpp]: Right - inliner will figure that out and optimize if that is still the case.
Not making assumptions about details hidden in the functions makes it more flexible to future changes.

---
In reply to: [106304921](https://github.com/Microsoft/ChakraCore/pull/2640#discussion_r106304921) [](ancestors = 106304921)

Comment (suwc) [inline lib/Runtime/Library/ModuleRoot.cpp]: Indeed - thanks :)

---
In reply to: [106304536](https://github.com/Microsoft/ChakraCore/pull/2640#discussion_r106304536) [](ancestors = 106304536)

File: lib/Runtime/Library/TypedArray.cpp (hunk 12 above)
Comment (suwc) [inline lib/Runtime/Library/TypedArray.cpp]: For typed arrays, the integer-indexed cases should flow through
BOOL TypedArrayBase::SetProperty(PropertyId propertyId, Var value, PropertyOperationFlags flags, PropertyValueInfo* info)
Seems like a good place for assertion + comment.

---
In reply to: [106301781](https://github.com/Microsoft/ChakraCore/pull/2640#discussion_r106301781) [](ancestors = 106301781)

File: lib/Runtime/Library/TypedArray.cpp (hunk 10 above)
Comment (suwc) [inline lib/Runtime/Library/TypedArray.cpp]: Being integer is guaranteed by IsNumericPropertyId(). Again I understand it may be confusing, but it has been around for long enough and entangled with names like "isNumeric" or "GetNumericValue", therefore I'd like to change them consistently, if at all, and probably get buy-in from 2/3 of the team. Leaving it alone for now.

---
In reply to: [106299416](https://github.com/Microsoft/ChakraCore/pull/2640#discussion_r106299416) [](ancestors = 106299416)

File: lib/Runtime/Library/TypedArray.cpp (hunk 11 above)
Comment (suwc) [inline lib/Runtime/Library/TypedArray.cpp]: var a = // some typed array
a[3];   // GetItemQuery
a[3]();   //  GetItemReferenceQuery


---
In reply to: [106301540](https://github.com/Microsoft/ChakraCore/pull/2640#discussion_r106301540) [](ancestors = 106301540)

**Comment by dilijev** [comment]
Thanks, Suwei. LGTM

**Comment by suwc** [comment]
Thank you all for reviewing this changeset.
